### PR TITLE
Upgrading terraform cloud in the UI using plan-only and allow-empty-apply runs 

### PR DESCRIPTION
**Pull request by brandonc**
### What

- Added new parameters to the run create API
- Added a section to Terraform Runs & Remote Operations / UI Driven Runs that details the upgrade process using plan-only and allow-empty-apply runs using the UI
- Added new modes to the run Modes and Options page

### Why

This is a new feature that is coming to Terraform Cloud

### Screenshots

Screen Shot 2022-05-11 at 11 30 37 AM

----------

### Merge Checklist
_If items do not apply to your changes, add (N/A) and mark them as complete._

#### Pull Request
- [x] One or more labels describe the type of change (e.g. clarification) and associated product (e.g. tfc).
- [x] Description links to related pull requests or issues, if any.

#### Content
- [ ] ~Redirects have been added for moved, renamed, or deleted pages.~
- [ ] ~Sidebar navigation files have been updated for added, deleted, reordered, or renamed pages.~
- [x] Links to related content where appropriate (e.g., API endpoints, permissions, etc.).
- [x] Pages with related content are updated and link to this content when appropriate.
- [ ] ~New pages have metadata (page name and description) at the top.~
- [x] New images are 2048 px wide. They have HashiCorp standard annotation color (#F92672) and format (rectangle with rounded corners), blurred sensitive details (e.g. credentials, usernames, user icons), and descriptive alt text in the markdown for accessibility.
- [ ] ~New code blocks have the correct syntax and line breaks to eliminate horizontal scroll bars.~
- [x] UI elements (button names, page names, etc.) are bolded.
- [ ] The Vercel website preview successfully deployed.

#### Reviews
- [ ] I or someone else reviewed the content for technical accuracy.
- [ ] I or someone else reviewed the content for typos, punctuation, spelling, and grammar.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1201832416281536

### CODE REVIEW

**Comment by laurapacilio** [comment]
Hi Brandon, is this the same as/related to this exissting PR from @sebasslash https://github.com/hashicorp/terraform-website/pull/2281/files

**Comment by brandonc** [comment]
@laurapacilio yes! I'll reconcile these. Thanks for pointing it out

File: content/cloud-docs/run/ui.mdx
Comment (sebasslash) [inline content/cloud-docs/run/ui.mdx:68]: Capital T (Terraform)

Comment (sebasslash) [inline content/cloud-docs/run/ui.mdx:68]:
```suggestion
Using the UI, you can start a new speculative plan with the workspace's current configuration version and any Terraform version available to the organization. After clicking **Start New Run**, use the select list to choose **Plan only** as the run type. This run mode reveals a new option to choose the specific terraform version.
```

**Review by sebasslash — CHANGES_REQUESTED** [review]
Overall, looking good 👍 Perhaps it might be worth mentioning the steps necessary to do a version jump like 0.11 ~> 0.13 ?

File: content/cloud-docs/api-docs/run.mdx
Comment (laurapacilio) [inline content/cloud-docs/api-docs/run.mdx:70]:
```suggestion
| `planned_and_finished` | The completion of a run containing a plan only, or a run that produces a plan with no changes to apply. This is a final state.                                                                                                                                                                                                                                                                                                    |
```

File: content/cloud-docs/api-docs/run.mdx
Comment (laurapacilio) [inline content/cloud-docs/api-docs/run.mdx:77]: Can we provide the tiniest bit more here. How is this different from just being canceled? Is it because of who is doing the canceling? or....what do we mean exactly? Also, should we say that this is a final state, like the others?

File: content/cloud-docs/api-docs/run.mdx
Comment (laurapacilio) [inline content/cloud-docs/api-docs/run.mdx:121]:
```suggestion
| `data.attributes.terraform-version`                | string               |                                                    | Specifies the terraform version to use in this run. Must be a valid terraform version available to the organization.                                                                                                                        |
```

File: content/cloud-docs/api-docs/run.mdx
Comment (laurapacilio) [inline content/cloud-docs/api-docs/run.mdx:122]: thoughts? Also + 1 to @sebasslash comment to do linking :) 

```suggestion
| `data.attributes.allow-empty-apply`                | bool                 |                                                    | Specifies whether Terraform can apply the run even when the plan contains no changes. Often used to upgrade state after upgrading a workspace to a new terraform version.                                          |
```

File: content/cloud-docs/run/modes-and-options.mdx
Comment (laurapacilio) [inline content/cloud-docs/run/modes-and-options.mdx:52]:
```suggestion
A no-operation (empty) apply enables Terraform Cloud to apply a run from a plan that contains no infrastructure changes. During apply, Terraform can upgrade the state version if required. You can use this option to upgrade the state in your Terraform Cloud workspace to a new Terraform version.
```

File: content/cloud-docs/run/modes-and-options.mdx
Comment (laurapacilio) [inline content/cloud-docs/run/modes-and-options.mdx:54]:
```suggestion
~> **Warning:** Terraform Cloud cannot guarantee that a plan in this mode will produce no changes. We recommend checking the plan for drift before proceeding to the apply stage. 
```

File: content/cloud-docs/run/modes-and-options.mdx
Comment (laurapacilio) [inline content/cloud-docs/run/modes-and-options.mdx:59]: I think we need to specify in this section that this is a speculative plan. Maybe we need to change the title to "Plan Only - Speculative Plan" or something similar. 

then we need to link to this new section you're making from this existing section on speculative plans: https://www.terraform.io/cloud-docs/run#speculative-plans. I recommend doing an in-text link. Something like : "Speculative plans are [Plan Only runs](/cloud-docs/run/modes-and-options). They show a set of possible changes and check them against Sentinel policies, but cannot apply those changes."

My concern here is just. We need to make sure that people understand that Plan Only = Speculative Plan. Thoughts?

File: content/cloud-docs/run/modes-and-options.mdx
Comment (laurapacilio) [inline content/cloud-docs/run/modes-and-options.mdx:61]:
```suggestion
The Plan Only option creates a plan that Terraform cannot apply. You can use it to test a specific terraform version that may differ from the workspace version. This lets you check whether your configuration is compatible with a newer terraform version without changing the workspace settings.
```

File: content/cloud-docs/run/modes-and-options.mdx
Comment (laurapacilio) [inline content/cloud-docs/run/modes-and-options.mdx:64]:
```suggestion
- **UI:** Open the workspace's **Actions** menu, select **Start new plan**, and then choose **Plan only** as the run type.
```

File: content/cloud-docs/run/ui.mdx
Comment (laurapacilio) [inline content/cloud-docs/run/ui.mdx:70]: My recommendation is to remove this screenshot. The reason why is that screenshots are difficult to maintain as our UI evolves, and I'm not sure they're necessary for contained dialog boxes like this. Can we try without?

File: content/cloud-docs/run/ui.mdx
Comment (laurapacilio) [reply content/cloud-docs/run/ui.mdx:68]: Can we also make the steps into a numbered list? Each item should start with a verb, and should go over one step. Be sure to tell folks what they should see when the step is completed (if appropriate). This is a best practice with ordered instructions and I think it will make this easier to read! If you take a first pass, I can edit for you :)

Comment (laurapacilio) [inline content/cloud-docs/run/ui.mdx:68]: Question -- I noticed you only added a section here on the Ui page that incorporates upgrading with speculative plans + empty apply. But I noticed that we do have an API attribute for empty apply now. Is this a process that folks could also do through the API? Is that recommended/something we expect folks to do? If so, should we also add a section about this on the API page? https://www.terraform.io/cloud-docs/run/api#advanced-use-cases

I'm just thinking that since we've added to the API, maybe we should provide folks with guidance about how to do this through that workflow :)

**Review by laurapacilio — CHANGES_REQUESTED** [review]
Hello! Thank you for working on this. I have some initial feedback/suggestions for you. Please let me know if you want to talk through them in zoom/you want more guidance. I am here to help!

- I think we need to add a new section called "Upgrading State" on this page: https://www.terraform.io/cloud-docs/workspaces/state.  I think there, we need to talk about _when_ folks are going to receive an error if their state is incompatible, the fact that Terraform automatically upgrades state on apply, the fact that you can use Empty apply to upgrade state (with a link! to the nice new section you made on run modes), etc. I also think we need to tell folks the process when they need to do this upgrading. I had read (I thought?) in a slack message that you can't upgrade from 0.10 to 0.12 in one go. You have to upgrade to 0.11 first, do your empty apply and then you can pick 0.12, do your empty apply again to finish the process. Is that true? There was some content in this from the other closed PR that I think we should bring back here and revise: https://www.terraform.io/cloud-docs/workspaces/state
- I think it could also be helpful on this page to ad one more subheading under "Plans and Applies" about "Allow Empty Apply" with a 2-3 sentence description of how this is used to update state, with a link to this section on Run Modes and Options (probably in text) and a link to the new section we're going to add on the state page where folks can find out more info.

File: content/cloud-docs/run/ui.mdx
Comment (nfagerlund) [reply content/cloud-docs/run/ui.mdx:68]: @laurapacilio This can certainly be done through the API as well; the UI even uses the API to do it, under the hood. But I'm feeling iffy about adding it to the API run workflow page, which has a very different vibe from this one and doesn't enumerate the various options in quite the same way.

File: content/cloud-docs/run/ui.mdx
Comment (nfagerlund) [inline content/cloud-docs/run/ui.mdx:66]: Hmm, how about,

```suggestion
## Testing Terraform Upgrades with Speculative Plans
```

**Review by nfagerlund — COMMENTED** [review]
I have a few wording suggestions. Other than that, I feel this is mostly on the right track, and defer to Laura on any remaining questions!

File: content/cloud-docs/run/ui.mdx
Comment (laurapacilio) [reply content/cloud-docs/run/ui.mdx:68]: This is a fair point, and maybe someday I'll rework the API page so that it does talk more about the options! Okay @brandonc -> I'm okay with leaving this out. But I wanted to bring it up just in case!

File: content/cloud-docs/api-docs/run.mdx
Comment (brandonc) [reply content/cloud-docs/api-docs/run.mdx:77]: I will add more. It means that the run could not be canceled gracefully and it was force-canceled by a workspace admin.

File: content/cloud-docs/run/modes-and-options.mdx
Comment (brandonc) [reply content/cloud-docs/run/modes-and-options.mdx:59]: Great idea. I was thinking that we should just stick with one phrase but upon further inspection Speculative plans are already documented extensively so it makes sense to connect these together.

**Comment by brandonc** [comment]
@laurapacilio @nfagerlund @sebasslash Great reviews! I'm so thankful for your help. I think I captured all your suggestions in the latest 3 commits

File: content/cloud-docs/run/ui.mdx
Comment (laurapacilio) [inline content/cloud-docs/run/ui.mdx:73]:
```suggestion
1. Click **Actions > Start New Run**. The **Start a new run** box appears.
1. Select **Plan only** as the run type. 
1. Select a version from the **Choose Terraform version** menu. The speculative plan will use this version without changing the workspace's settings.
1. Click **Start run**.

If the run is successful, you can change the workspace's Terraform version and [upgrade the state](/cloud-docs/workspaces/state#upgrading-state).
```

**Review by laurapacilio — CHANGES_REQUESTED** [review]
Did a more thorough review of language :) I like this direction and I think we're close! I'd like to take one more final pass after you sift through these proposed edits. As always, push back on me if my edits make things incorrect/you think something is missing!

Thank you for your great work on this!

File: content/cloud-docs/api-docs/run.mdx
Comment (nfagerlund) [reply content/cloud-docs/api-docs/run.mdx:113]: I think I'm fine either way on this one, but I'll note in passing that "destroy plan" _is_ the in-UI name of this operation type, and we say it verbatim in places like the run details page, the runs list, and the page where you can kick one off.

File: content/cloud-docs/run/ui.mdx
Comment (nfagerlund) [reply content/cloud-docs/run/ui.mdx:73]: Hmm, I think these instructions wandered off course a little bit, both in the current diff and Laura's suggestion. 

It looks like we're trying to outline TWO distinct but related procedures here: 

- Do a plan-only run with a different Terraform version to find out whether it'll work OK or not.
- Actually flip the workspace Terraform version...
    - ...and, _maybe,_ do an allow empty apply run to lock in a state version upgrade. (Only necessary on certain version transitions, most notably when landing on 0.13.)

I think the instructions will be more clear if they treat those two actions as distinct.

Comment (laurapacilio) [reply content/cloud-docs/run/ui.mdx:73]: I agree that it's two separate things! But I thought (let me know if I'm super off base here) that this edit really outlines the process for the first procedure, which is what this section is about. This is essentially how you test, right? You run the speculative (plan-only) run with the version you want selected and see if it's successful? And since this is called "Testing Terraform Upgrades with Speculative Plans", I was thinking it makes more sense to really only outline the procedure to do that.

And then my thought is that you can basically say "cool! your test was successful? Do you wanna actually change the version now and/or do the state migration? Here are links to that. This is what that last sentence was attempting to do.

That said, please let me know if I'm way off base here --> I just wanted to provide some context for this edit. It may be that there are gaps in my knowledge/that my execution of this was unclear. Anyway, I value your opinion so let me know what you think!

Comment (nfagerlund) [reply content/cloud-docs/run/ui.mdx:73]: @laurapacilio You know, I'm not sure where my head was at on Friday, but on re-reading your suggested text, I agree with your edit! Possibly I got confused about the state of the header. Comment retracted.

Comment (laurapacilio) [reply content/cloud-docs/run/ui.mdx:73]: okay! I still appreciate you and your feedback! :)

**Comment by brandonc** [comment]
@laurapacilio Thanks for the detailed review. I think this is much better than where it began!

**Review by laurapacilio — APPROVED** [review]
Approving so I don't block and you can get this in before today's migration! Thank you for your work on this, I believe it's much improved from the first draft and I really really appreciate the opportunity to comment and help! :)

**Comment by thiskevinwang** [comment]
Hey @brandonc, the Vercel deploy preview is failing due to an expected error on the `/downloads` page.

If you merge in or rebase onto the latest `master`, the preview should deploy :)

**Pull request merged**
